### PR TITLE
Add support/BytesToHex.h -> lightweight hex conversion

### PR DESCRIPTION
**Pull request by tcarmelveilleux**
#### Problem
PR https://github.com/project-chip/connectedhomeip/pull/4362 and future operational mDNS client/server usage require hex encoded strings in uppercase/lowercase. Some manual code exists to do so, without tests.

 #### Summary of Changes
Add a lightweight hex conversion function and associated tests.

- Lean on stack
- Supports safe output to buffers of any size
- Support uppercase/lowercase
- Support non-nul-terminated and nul-terminated

### CODE REVIEW

**Review by mspang — APPROVED** [review]

File: src/lib/support/BytesToHex.cpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        return x + '0';
+    }
+}
+
+} // namespace
+
+CHIP_ERROR BytesToHex(const uint8_t * src_bytes, size_t src_size,
+                      char * dest_hex, size_t dest_size_max,
+                      bool uppercase, bool nul_terminate)
```
Comment (andy31415) [inline src/lib/support/BytesToHex.cpp]: consider using flags. Consecutive boolean calls are not easy to read/check (i.e. prefer `BytesToHex(src, srcSize, dest, destSize,HexFlags::kUppercase | HexFlags::kNullTerminate)` to `BytesToHex(src, srcSize, dest, destSize, true, true)`

**Review by andy31415 — APPROVED** [review]

**Review by saurabhst — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

File: src/lib/support/BytesToHex.cpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+CHIP_ERROR BytesToHex(const uint8_t * src_bytes, size_t src_size, char * dest_hex, size_t dest_size_max, bool uppercase,
+                      bool nul_terminate)
+{
+    if ((src_bytes == nullptr) || (dest_hex == nullptr))
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    size_t expected_output_size = (src_size * 2u) + (nul_terminate ? 1u : 0u);
```
Comment (bzbarsky-apple) [inline src/lib/support/BytesToHex.cpp:49]: Might be worth a comment about why we're note worried about overflow here.

Or maybe we _should_ worry about it, and should rewrite the test accordingly:
```
if (nul_terminate) {
  if (dest_size_max / 2 < expected_output_size || 2*expected_output_size >= dest_size_max) {
    return CHIP_ERROR_BUFFER_TOO_SMALL;
  }
} else {
  if (dest_size_max / 2 < expected_output_size || 2*expected_output_size > dest_size_max) {
    return CHIP_ERROR_BUFFER_TOO_SMALL;
  }
}
```
or some such; I am not a big fan of the comprehensibility of that code....

File: src/lib/support/BytesToHex.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    if (dest_size_max < expected_output_size)
+    {
+        return CHIP_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    char * cursor = dest_hex;
+    for (size_t byte_idx = 0; byte_idx < src_size; ++byte_idx)
+    {
+        *cursor++ = NibbleToHex((src_bytes[byte_idx] >> 4) & 0xFu, uppercase);
```
Comment (bzbarsky-apple) [inline src/lib/support/BytesToHex.cpp:58]: Why do we need to mask with `0xF` both here and inside `NibbleToHex`?  Even ignoring for the moment that after shifting a `uint8_t` right by 4 we better end up with no bits set outside `0xF`.

File: src/lib/support/BytesToHex.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        *cursor++ = NibbleToHex((src_bytes[byte_idx] >> 0) & 0xFu, uppercase);
```
Comment (bzbarsky-apple) [inline src/lib/support/BytesToHex.cpp:59]: Why the right-shift by 0 here?  Just for symmetry?

File: src/lib/support/BytesToHex.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+#include <core/CHIPError.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+namespace chip {
+namespace Encoding {
+
+/**
+ * Encode a buffer of bytes into hexadecimal, with or without nul-termination
```
Comment (bzbarsky-apple) [inline src/lib/support/BytesToHex.h:28]: I think usually one would write "null", not "nul", fwiw...

File: src/lib/support/BytesToHex.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+ * The function returns CHIP_ERROR_BUFFER_TOO_SMALL if dest_size_max can't fit
+ * the entire encoded buffer, and the nul-terminator if enabled. This function
+ * will never output truncated data. The result either fits and is written,
+ * or does not fit and nothing is written to dest_hex.
+ *
+ * If src_bytes and dest_hex overlap, the results may be incorrect, depending
+ * on overlap, but only the core validity checks are done and it's possible to
+ * get CHIP_NO_ERROR with erroneous output.
+ *
+ * On success, number of writes written to destination is always:
```
Comment (bzbarsky-apple) [inline src/lib/support/BytesToHex.h:41]:
```suggestion
 * On success, number of bytes written to destination is always:
```

File: src/lib/support/BytesToHex.h [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+ *   output_size = (src_size * 2) + (nul_terminate ? 1 : 0);
+ *
+ * @param src_bytes Pointer to non-null buffer to convert
+ * @param src_size Number of bytes to convert from src_bytes
+ * @param [out] dest_hex Destination buffer to receive hex encoding
+ * @param dest_size_max Maximum buffer size for the hex encoded `dest_hex` buffer
+ *                      including nul-terminator if needed.
+ * @param uppercase Use upper-case hex if true, otherwise lower-case hex
+ * @param nul_terminate Add nul terminator if true. See description
```
Comment (bzbarsky-apple) [inline src/lib/support/BytesToHex.h:50]: Fwiw, using enum classes for those args would make callsites clearer without requiring people to add comments explaining what the booleans mean, and make it harder to get the two booleans mixed up by accident.  I would strongly urge we do that.

**Review by bzbarsky-apple — APPROVED** [review]
We can use this today, in DiscoveryImplPlatform::SetupHostname.

Is it worth creating some template overloads that auto-detect sizes from the size of the stack buffer you're passing in, for consumers like that one that have stack buffers?

File: src/lib/support/BytesToHex.h (hunk 7 above)
Comment (tcarmelveilleux) [reply src/lib/support/BytesToHex.h:50]: Yes, I should have done that right off the bat. Done now.

File: src/lib/support/BytesToHex.cpp (hunk 4 above)
Comment (tcarmelveilleux) [reply src/lib/support/BytesToHex.cpp:59]: Yes, just for symmetry

File: src/lib/support/BytesToHex.cpp (hunk 3 above)
Comment (tcarmelveilleux) [reply src/lib/support/BytesToHex.cpp:58]: Symmetry and not relying on future impls of NibbleToHex. I can remove. BTW, this also will be optimized out.

File: src/lib/support/BytesToHex.cpp (hunk 2 above)
Comment (tcarmelveilleux) [reply src/lib/support/BytesToHex.cpp:49]: Found a clean way

File: src/lib/support/BytesToHex.cpp (hunk 1 above)
Comment (tcarmelveilleux) [reply src/lib/support/BytesToHex.cpp]: Done

File: src/lib/support/BytesToHex.cpp (hunk 3 above)
Comment (tcarmelveilleux) [reply src/lib/support/BytesToHex.cpp:66]: Forgot to remove on my review addressing commit :(. I think it's fine to leave, but if it's very important I will update.

**Pull request merged**